File: OpenMP/EPCC/Fortran/schedbench.f90
!***************************************************************************
!*                                                                         *
!*             OpenMP MicroBenchmark Suite - Version 2.0                   *
!*                                                                         *
!*                            produced by                                  *
!*                                                                         *
!*                     Mark Bull and Fiona Reid                            *
!*                                                                         *
!*                                at                                       *
!*                                                                         *
!*                Edinburgh Parallel Computing Centre                      *
!*                                                                         *
!*         email: markb@epcc.ed.ac.uk or fiona@epcc.ed.ac.uk               *
!*                                                                         *
!*                                                                         *
!*      This version copyright (c) The University of Edinburgh, 2004.      *
!*                         All rights reserved.                            *
!*                                                                         *
!***************************************************************************

program schedbench

  use benchdata
  implicit none

! GET NUMBER OF THREADS

!$OMP PARALLEL 
!$OMP MASTER
  nthreads=omp_get_num_threads()
!$OMP END MASTER 
!$OMP END PARALLEL 
  print *,"Running OpenMP scheduling benchmark on ",nthreads,&
  " thread(s)"

! FIX LOOP BODY LENGTH 

  call getdelay() 

! GENERATE REFERENCE TIME 

  call refer() 

! TEST STATIC 

  call teststatic()

! TEST STATIC,n 
 
  cksz = 1
  do while (cksz .le. itersperthr)
    call teststaticn() 
    cksz = cksz * 2 
  end do 

! TEST DYNAMIC,n 

  cksz = 1 
  do while (cksz .le. itersperthr)
    call testdynamicn() 
    cksz = cksz * 2 
  end do 

! TEST GUIDED,n 

  cksz = 1 
  do while (cksz .le. itersperthr/nthreads)
    call testguidedn() 
    cksz = cksz * 2 
  end do 

  stop
end program schedbench

subroutine teststatic()

  use benchdata
  use benchstats
  implicit none

  integer :: i, j, k, dl
  real (kind = dp) :: start
  real (kind = dp) :: meantime, sd 
  real (kind = dp) :: getclock 

  write(6,*) 
  write(6,1010)
  write(6,1100)
  1010 format("--------------------------------------------------------")
  1100 format("Computing STATIC time") 
      
  dl = delaylength 

  do k=0,outerreps

! PARALLEL TIME 

    start  = getclock()
!$OMP PARALLEL PRIVATE(J) 
      do j=1,innerreps_sched
!$OMP DO SCHEDULE(STATIC) 
        do i=1,itersperthr*nthreads
          call delay(dl)
        end do
      end do
!$OMP END PARALLEL
      time(k) = (getclock() - start) * 1.0e6 / dble (innerreps_sched) 
    end do

! PROCESS TIMING DATA 

  call stats (meantime, sd) 

  write (6,1000) meantime, conf95*sd
  1000 format("STATIC time =                         ",f10.2,&
      " microseconds +/- ", f8.3)
  write (6,*) 
  write (6,1001) meantime-reftime, conf95*(sd+refsd)
  1001 format("STATIC overhead =                   ",f10.2,&
      " microseconds +/- ", f8.3)
  return

end subroutine teststatic

subroutine teststaticn()

  use benchdata
  use benchstats
  implicit none

  integer :: i, j, k, dl, mycksz 
!  mycksz - horrid workaround for Guide 3.7 bug! 
  real (kind = dp) :: start
  real (kind = dp) :: meantime, sd
  real (kind = dp) :: getclock 

  write(6,*) 
  write(6,1010)
  write(6,1100) cksz 
  1010 format("--------------------------------------------------------")
  1100 format("Computing STATIC",i5," time") 

  dl = delaylength 
      
  do k=0,outerreps

! PARALLEL TIME 

    start  = getclock()
!$OMP PARALLEL PRIVATE (j,mycksz) 
    mycksz = cksz  
    do j=1,innerreps_sched
!$OMP DO SCHEDULE(STATIC,mycksz) 
      do i=1,itersperthr*nthreads
        call delay(dl)
      end do
    end do
!$OMP END PARALLEL
    time(k) = (getclock() - start) * 1.0e6 / dble (innerreps_sched) 
  end do

! PROCESS TIMING DATA 

  call stats (meantime, sd) 

  write (6,1000) cksz, meantime, conf95*sd
  1000 format("STATIC",i5," time =                       ",f10.2,&
      " microseconds +/- ", f8.3)
  write (6,*) 
  write (6,1001) cksz, meantime-reftime, conf95*(sd+refsd)
  1001 format("STATIC",i5," overhead =                   ",f10.2,&
      " microseconds +/- ", f8.3)
  return

end subroutine teststaticn

subroutine testdynamicn()

  use benchdata
  use benchstats
  implicit none

  integer :: i, j, k, dl, mycksz 
!  mycksz - horrid workaround for Guide 3.7 bug! 
  real (kind = dp) :: start
  real (kind = dp) :: meantime, sd
  real (kind = dp) :: getclock 

  write(6,*) 
  write(6,1010)
  write(6,1100) cksz 
  1010 format("--------------------------------------------------------")
  1100 format("Computing DYNAMIC",i5," time") 
       
  dl = delaylength 

  do k=0,outerreps

! PARALLEL TIME 

    start  = getclock()
!$OMP PARALLEL PRIVATE (j, mycksz)
    mycksz = cksz  
    do j=1,innerreps_sched
!$OMP DO SCHEDULE(DYNAMIC,mycksz) 
      do i=1,itersperthr*nthreads
        call delay(dl)
      end do
    end do
!$OMP END PARALLEL
    time(k) = (getclock() - start) * 1.0e6 / dble (innerreps_sched) 
  end do

! PROCESS TIMING DATA 

  call stats (meantime, sd) 

  write (6,1000) cksz, meantime, conf95*sd
  1000 format("DYNAMIC",i5," time =                       ",f10.2,&
      " microseconds +/- ", f8.3)
  write (6,*) 
  write (6,1001) cksz, meantime-reftime, conf95*(sd+refsd)
  1001 format("DYNAMIC",i5," overhead =                   ",f10.2,&
      " microseconds +/- ", f8.3)
  return

end subroutine testdynamicn

subroutine testguidedn()

  use benchdata
  use benchstats
  implicit none

  integer :: i, j, k, dl, mycksz 
!  mycksz - horrid workaround for Guide 3.7 bug! 
  real (kind = dp) :: start
  real (kind = dp) :: meantime, sd
  real (kind = dp) :: getclock 

  write(6,*) 
  write(6,1010)
  write(6,1100) cksz 
  1010 format("--------------------------------------------------------")
  1100 format("Computing GUIDED",i5," time") 

  dl = delaylength 
      
  do k=0,outerreps

! PARALLEL TIME 

    start  = getclock()
!$OMP PARALLEL PRIVATE (j,mycksz)
    mycksz = cksz 
    do j=1,innerreps_sched
!$OMP DO SCHEDULE(GUIDED,mycksz) 
      do i=1,itersperthr*nthreads
        call delay(dl)
      end do
    end do
!$OMP END PARALLEL
    time(k) = (getclock() - start) * 1.0e6 / real (innerreps_sched) 
  end do

! PROCESS TIMING DATA 

  call stats (meantime, sd) 

  write (6,1000) cksz, meantime, conf95*sd
  1000 format("GUIDED",i5," time =                       ",f10.2,&
      " microseconds +/- ", f8.3)
  write (6,*) 
  write (6,1001) cksz, meantime-reftime, conf95*(sd+refsd)
  1001 format("GUIDED",i5," overhead =                   ",f10.2,&
      " microseconds +/- ", f8.3)
  return

end subroutine testguidedn

subroutine refer() 

  use benchdata
  use benchstats
  implicit none

  integer :: i, j, k, dl
  real (kind = dp) :: start
  real (kind = dp) :: meantime, sd
  real (kind = dp) :: getclock 

  write(6,*) 
  write(6,1010)
  write(6,1100)
  1010 format("--------------------------------------------------------")
  1100 format("Computing reference time ") 

! SEQUENTIAL REFERENCE TIME 

  dl = delaylength 

  do k=0,outerreps
    start  = getclock()
    do j=1,innerreps_sched
      do i=1,itersperthr
        call delay(dl)
      end do
    end do
    time(k) = (getclock() - start)* 1.0e6 / real (innerreps_sched) 
  end do

! PROCESS TIMING DATA 

  call stats (meantime, sd) 

  reftime = meantime 
  refsd = sd 
  write (6,1000) reftime, conf95*refsd
  1000 format("Reference_time =                      ",f10.2,&
  " microseconds +/- ", f8.3)
  write (6,*) 
 return

end subroutine refer

subroutine getdelay() 

  use benchdata
  use benchstats
  implicit none

  integer :: i, reps, dl  
  real (kind = dp) :: actualtime, targettime
  real (kind = dp) :: start
  real (kind = dp) :: getclock   

! CHOOSE delaylength SO THAT call delay(delaylength) 
! TAKES APPROXIMATELY 100 CPU CLOCK CYCLES
 
  dl = 0
  reps = itersperthr * innerreps_sched 
  actualtime = 0. 
  targettime = 100.0 / mhz * 1e-06 

  call delay(dl) 

  do while (actualtime .lt. targettime) 
    dl = dl * 1.1 + 1 
    start = getclock()
      do i = 1, reps  
        call delay(dl) 
      end do 
      actualtime  = (getclock() - start) / dble  (reps)
  end do
  
  start = getclock()
  do i = 1, reps  
    call delay(dl) 
  end do 
  actualtime  = (getclock() - start) / dble (reps)
  delaylength = dl 
  print *, "Assumed clock rate = ",mhz," MHz "
  print *, "Delay length = ", dl
  print *, "Delay time  = ", actualtime * mhz * 1e6, "  cycles"
  return

end subroutine getdelay

File: OpenMP/EPCC/C/schedbench.c
/****************************************************************************
*                                                                           *
*             OpenMP MicroBenchmark Suite - Version 4.0                     *
*                                                                           *
*                            produced by                                    *
*                                                                           *
*                             Mark Bull                                     *
*                                                                           *
*                                at                                         *
*                                                                           *
*                   EPCC, University of Edinburgh                           *
*                                                                           *
*                    email: m.bull@epcc.ed.ac.uk                            *
*                                                                           *
*                                                                           *
*      This version copyright (c) The University of Edinburgh, 2023.        *
*                                                                           *
*                                                                           *
*  Licensed under the Apache License, Version 2.0 (the "License");          *
*  you may not use this file except in compliance with the License.         *
*  You may obtain a copy of the License at                                  *
*                                                                           *
*      http://www.apache.org/licenses/LICENSE-2.0                           *
*                                                                           *
*  Unless required by applicable law or agreed to in writing, software      *
*  distributed under the License is distributed on an "AS IS" BASIS,        *
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
*  See the License for the specific language governing permissions and      *
*  limitations under the License.                                           *
*                                                                           *
****************************************************************************/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <omp.h>

#include "common.h"
#include "schedbench.h"

int cksz, itersperthr = 1024;
char testName[32];
extern char type[120];
int main(int argc, char **argv) {

    init(argc, argv);
    // Also print itersperthr
    printf("\t%d iterations per thread\n", itersperthr);

    /* GENERATE REFERENCE TIME */
    reference("reference time", &refer);

    /* TEST STATIC */
    if((strcmp("STATIC",type)==0)||(strcmp("ALL",type)==0)){
      benchmark("STATIC", &teststatic);
    }

    if((strcmp("STATIC_MONOTONIC",type)==0)||(strcmp("ALL",type)==0)){
      benchmark("STATIC_MONOTONIC", &teststaticmono);
    }



    /* TEST STATIC,n */
    if((strcmp("STATICN",type)==0)||(strcmp("ALL",type)==0)){
      cksz = 1;
      while (cksz <= itersperthr) {
        sprintf(testName, "STATIC %d", cksz);
        benchmark(testName, &teststaticn);
        cksz *= 2;
      }
    }

    if((strcmp("STATICN_MONOTONIC",type)==0)||(strcmp("ALL",type)==0)){
      cksz = 1;
      while (cksz <= itersperthr) {
        sprintf(testName, "STATIC_MONOTONIC %d", cksz);
        benchmark(testName, &teststaticnmono);
        cksz *= 2;
      }
    }


    /* TEST DYNAMIC,n */
    if((strcmp("DYNAMIC",type)==0)||(strcmp("ALL",type)==0)){
      cksz = 1;
      while (cksz <= itersperthr) {
        sprintf(testName, "DYNAMIC %d", cksz);
        benchmark(testName, &testdynamicn);
        cksz *= 2;
      }
    }

    if((strcmp("DYNAMIC_MONOTONIC",type)==0)||(strcmp("ALL",type)==0)){
      cksz = 1;
      while (cksz <= itersperthr) {
        sprintf(testName, "DYNAMIC_MONOTONIC %d", cksz);
        benchmark(testName, &testdynamicnmono);
        cksz *= 2;
      }
    }

    /* TEST GUIDED,n */
    if((strcmp("GUIDED",type)==0)||(strcmp("ALL",type)==0)){
      cksz = 1;
      while (cksz <= itersperthr / nthreads) {
        sprintf(testName, "GUIDED %d", cksz);
        benchmark(testName, &testguidedn);
      cksz *= 2;
      }
    }

    if((strcmp("GUIDED_MONOTONIC",type)==0)||(strcmp("ALL",type)==0)){
      cksz = 1;
      while (cksz <= itersperthr / nthreads) {
        sprintf(testName, "GUIDED_MONOTONIC %d", cksz);
        benchmark(testName, &testguidednmono);
        cksz *= 2;
      }
    }

    if((strcmp("TASKLOOP",type)==0)||(strcmp("ALL",type)==0)){
      cksz = 1;
      while (cksz <= itersperthr / nthreads) {
        sprintf(testName, "TASKLOOP %d", cksz);
        benchmark(testName, &taskloopn);
        cksz *= 2;
      }
    }

    finalise();
    return EXIT_SUCCESS;

  }

void refer() {
    int i, j;
    for (j = 0; j < innerreps; j++) {
      for (i = 0; i < itersperthr; i++) {
        delay(delaylength);
      }
    }
}

void teststatic() {

    int i, j;
#pragma omp parallel private(j)
    {
	for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(static)
	    for (i = 0; i < itersperthr * nthreads; i++) {
		delay(delaylength);
	    }
	}
    }
}
void teststaticmono() {

    int i, j;
#pragma omp parallel private(j)
    {
	for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(monotonic:static)
	    for (i = 0; i < itersperthr * nthreads; i++) {
		delay(delaylength);
	    }
	}
    }
}

void teststaticn() {
    int i, j;
#pragma omp parallel private(j)
    {
	for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(static,cksz)
	    for (i = 0; i < itersperthr * nthreads; i++) {
		delay(delaylength);
	    }
	}
    }
}

void teststaticnmono() {
    int i, j;
#pragma omp parallel private(j)
    {
    for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(monotonic:static,cksz)
        for (i = 0; i < itersperthr * nthreads; i++) {
        delay(delaylength);
        }
      }
    }
}


void testdynamicn() {
    int i, j;
#pragma omp parallel private(j)
    {
    for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(dynamic,cksz)
        for (i = 0; i < itersperthr * nthreads; i++) {
        delay(delaylength);
        }
      }
    }
}

void testdynamicnmono() {
    int i, j;
#pragma omp parallel private(j)
    {
    for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(monotonic:dynamic,cksz)
        for (i = 0; i < itersperthr * nthreads; i++) {
        delay(delaylength);
        }
    }
    }
}

void testguidedn() {
    int i, j;
#pragma omp parallel private(j)
    {
    for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(guided,cksz)
        for (i = 0; i < itersperthr * nthreads; i++) {
        delay(delaylength);
        }
      }
    }
}
void testguidednmono() {
    int i, j;
#pragma omp parallel private(j)
    {
    for (j = 0; j < innerreps; j++) {
#pragma omp for schedule(monotonic:guided,cksz)
        for (i = 0; i < itersperthr * nthreads; i++) {
        delay(delaylength);
        }
      }
    }
}

void taskloopn() {
  int i, j;
#pragma omp parallel private(j)
  {
    #pragma omp master
    {
      for (j = 0; j < innerreps; j++) {
        #pragma omp taskloop num_tasks((itersperthr*nthreads)/cksz)
        for (i = 0; i < itersperthr * nthreads; i++) {
          delay(delaylength);
        }
      }
    }
  }
}
